### PR TITLE
Adds `ui-date-range-picker` and refactors `ui-date-picker`

### PR DESCRIPTION
**Pull request by pzuraq**
Removes the `day-picker-day` component in favor of simple ember objects for performance reasons. Adds the range picker components, comments, and tests.

### CODE REVIEW

File: tests/integration/date-pickers/ui-date-picker-test.js [hunk 1]
```diff
@@ -69,39 +49,95 @@ test('ui-date-picker component', function(assert) {
...
     }).then(() => {
-      assert.ok(find('.popup-menu').is(':visible') === true, 'Calendar popup appears when the button is clicked');
+      assert.ok(find(`${datePickerId} ${dropdownClass}`).is(':visible'), 'Calendar popup appears when the button is clicked');
 
-      let today = new Date();
-      let displayedDate = monthNames[today.getMonth()].fullName + ' ' + today.getFullYear();
-      assert.ok(find('.tf-date-picker__current-month').text() === displayedDate, 'Current month and year on the calendar popup are correct');
+      assert.equal(find(`${datePickerId} ${currentMonthClass}`).text(), today.format('MMMM YYYY'), 'Current month and year on the calendar popup are correct');
 
       // click the first of the month
```
Comment (dinomiike) [inline tests/integration/date-pickers/ui-date-picker-test.js]: if you're going to change to date it clicks to the 11th, this comment should be removed. also the name firstDaySelector implies the first day clicked? given the context of this change i wasn't sure if that's what it was mean to be.

Comment (pzuraq) [inline tests/integration/date-pickers/ui-date-picker-test.js]: firstDaySelector still works for the time being as it's the first day to be selected, though if you can think of a better name I'm all for it! Comment should be removed, will do that soon. Also the change to the 11th/13th was to guarantee a single date would be selected (it was impossible to select the 2nd with just CSS selectors.

File: addon/components/ui-date-picker.js [hunk 2]
```diff
@@ -3,96 +3,194 @@ import moment from 'moment';
...
+   * Day is the selected day in `datePicker`
+   *
+   * @property isSelected
+   */
+  isSelected: computed('value', function() {
+    return this.get('date').isSame(this.get('value') || null, 'day');
+  }),
+
+  /**
+   * Day is disabled
```
Comment (null-null-null) [inline addon/components/ui-date-picker.js]: Is this because you're only displaying one month at a time? Either way, would be useful to know exactly why this is the condition for disabling the day.

Comment (pzuraq) [inline addon/components/ui-date-picker.js]: This currently only applies a style to the day, there is no way to actually disable days yet. Coming in another PR.

File: addon/components/ui-date-range-picker.js [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+  set,
+  computed,
+  observer
+} = Ember;
+
+const {
+  alias,
+  reads,
+  or
+} = Ember.computed;
```
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: `} = computed;`

File: addon/components/ui-date-range-picker.js [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+  rangeIsLastWeek: rangeIs('startOfLastWeek', 'endOfLastWeek'),
+  rangeIsThisMonth: rangeIs('startOfThisMonth', 'today'),
+  rangeIsLastMonth: rangeIs('startOfLastMonth', 'endOfLastMonth'),
+  rangeIsThisQuarter: rangeIs('startOfThisQuarter', 'today'),
+  rangeIsLastQuarter: rangeIs('startOfLastQuarter', 'endOfLastQuarter'),
+  rangeIsThisYear: rangeIs('startOfThisYear', 'today'),
+  rangeIsLastYear: rangeIs('startOfLastYear', 'endOfLastYear')
+});
+
+function rangeIs(start, end) {
```
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: Should be put into its own file so it can be tested. `app/computeds/range.js`, or something to that effect.

Comment (pzuraq) [inline addon/components/ui-date-range-picker.js]: Since this computed is not currently being used elsewhere, and likely won't be due to it's specificity, I decided it was best not to put it into the computed's namespace. If we find a need for it elsewhere's we can pull it out.

File: addon/components/ui-date-range-picker.js [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+   * Returns the currently selected period in months or days, for changing periods
+   * directly.
+   *
+   * @property period
+   */
+  period: computed('startDate', 'endDate', function() {
+    const startDate = moment(this.get('startDate') || null);
+    const endDate = moment(this.get('endDate') || null);
+
+    const periodType = this.get('rangeIsInMonths')   ? 'months' : 'days';
```
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: There are 3 spaces here: `this.get('rangeIsInMonths')   ?`. Should probably be 1.

File: addon/components/ui-date-range-picker.js (hunk 4 above)
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: In this case it's less about reuse and more about testing.

File: addon/templates/components/ui-date-picker.hbs [hunk 6]
```diff
@@ -18,27 +18,25 @@
...
     <thead>
       <tr>
-          <th class="tf-date-picker__calendar-header">S</th>
-          <th class="tf-date-picker__calendar-header">M</th>
-          <th class="tf-date-picker__calendar-header">T</th>
-          <th class="tf-date-picker__calendar-header">W</th>
-          <th class="tf-date-picker__calendar-header">T</th>
-          <th class="tf-date-picker__calendar-header">F</th>
-          <th class="tf-date-picker__calendar-header">S</th>
+        {{#each dayOfWeek in dayNames}}
```
Comment (null-null-null) [inline addon/templates/components/ui-date-picker.hbs]: `{{#each dayNames as |dayOfWeek|}}`

File: addon/components/ui-date-range-picker.js (hunk 4 above)
Comment (pzuraq) [inline addon/components/ui-date-range-picker.js]: The benefits of adding testing to one computed are outweighed by the negative of adding it to the global namespace.

File: addon/templates/components/ui-date-range-picker.hbs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      </a>
+
+      <div class="tf-date-picker__current-month">{{moment currentMonth "MMMM YYYY"}}</div>
+
+    </header>
+
+    <table class="tf-date-picker__calendar-month">
+      <thead>
+        <tr>
+          {{#each dayOfWeek in dayNames}}
```
Comment (null-null-null) [inline addon/templates/components/ui-date-range-picker.hbs]: {{#each dayNames as |dayOfWeek|}}

File: addon/templates/components/ui-date-range-picker.hbs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            <th class="tf-date-picker__calendar-header">{{dayOfWeek}}</th>
+          {{/each}}
+        </tr>
+      </thead>
+      <tbody>
+        {{#each week in weeks}}
+          <tr>
```
Comment (null-null-null) [inline addon/templates/components/ui-date-range-picker.hbs]: {{#each weeks as |week|}}

File: addon/templates/components/ui-date-range-picker.hbs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            {{#each day in week}}
```
Comment (null-null-null) [inline addon/templates/components/ui-date-range-picker.hbs]: {{#each week as |day|}}

File: addon/components/ui-date-range-picker.js (hunk 4 above)
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: Are you adding it to the global namespace if you put it in a module?

Comment (pzuraq) [inline addon/components/ui-date-range-picker.js]: If you place the computed in `app/computeds/range.js` or `addon/computeds/range.js` it will be accessible from pretty much anywhere. Even if we put it in its own folder it will still be accessible.

When pod structure becomes more finalized we may be able to pull it out. For now we can test this computed by actually testing the date-picker component's behavior itself. I'll add a check to the unit tests.

**Comment by null-null-null** [comment]
Solid approach, helpful comments, and thorough tests. Just a few small style points. Also, looks like the PR can't yet be merged due to conflicts.

File: addon/components/ui-date-range-picker.js (hunk 4 above)
Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: If you encapsulate it in a module, it isn't technically global, right?
Since you'll have to import it into other files in order to access it. An
example
<https://github.com/stefanpenner/ember-moment/blob/master/addon/computeds/ago.js>
from ember-moment.

Comment (pzuraq) [inline addon/components/ui-date-range-picker.js]: Yes, sorry I think we're using different meanings for global. I'm not talking about the global variable namespace, I'm talking about the Ember module namespace. You're still making a computed that can be pulled in anywhere, even places we don't want it to be.

It's basically making a private API public, but not telling people and hoping they won't use it.

Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: I'm not attached to this approach (and it's convenient to see the code for `rangeIs` right there in the same file), but this is in the [styleguide](https://github.com/Ticketfly/javascript-styleguide#unit-tests). Might be worth the future benefit of testing in isolation.

Comment (pzuraq) [inline addon/components/ui-date-range-picker.js]: "Make sure to extract __complicated__ computed properties"

Comment (null-null-null) [inline addon/components/ui-date-range-picker.js]: Good enough for me.
